sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -76,6 +77,13 @@ case class Project(projectList: Seq[NamedExpression], child: LogicalPlan)
 
   override lazy val validConstraints: ExpressionSet =
     getAllValidConstraints(projectList)
+
+  val hiddenOutputTag: TreeNodeTag[Seq[Attribute]] = TreeNodeTag[Seq[Attribute]]("hiddenOutput")
+
+  override def metadataOutput: Seq[Attribute] = {
+    child.metadataOutput ++
+      getTagValue(hiddenOutputTag).getOrElse(Seq.empty[Attribute])
```

Comment:
It's unfortunate that we need to use `TreeNodeTag` to store the extra information in `Project`, but I don't have a better idea without changing the `Project` constructor.

Reply:
We could make this more generic by adding this `LogicalPlan`'s `metadataOutput`, but that would complicate how we can add these hidden columns in `AddMetadataColumns`.